### DISCUSSION TITLE
Need help understanding events in mqtt_as.py

### DISCUSSION THREAD

**Discussion by tferrin** [micropython/micropython#12466] [Libraries & Drivers] [answered]
I'm a newbie when it comes to using the event interface in mqtt_as.py. I've tried the range_ex.py example in the mqtt_as distribution and I've cookie-cuttered code from that example into my own code and both work great. But there's a code construct that I just don't understand:

```
from mqtt_as import MQTTClient
client = MQTTClient(config)

async def up(client):
    while True:
        await client.up.wait()
        client.up.clear()
        wifi_led(True)
        print('We are connected to broker.')
        await client.subscribe(sensorTopic, 1)
```
How is this legal Python? MQTTClient doesn't have an "up" method (or "down" method either), so how can the line "await client.up.wait()" not raise an error? I know for this code to work it requires "config["queue_len"] = 4" but I don't understand how this this apparently changes the semantics of the statement "client.up.wait()". Can someone please point me to documentation that explains what's going on. Thanks!

**Comment by andrewleech** [comment, marked as answer]
`up` and `down` here are `asyncio.Event()` objects, see https://github.com/peterhinch/micropython-mqtt/blob/f9290c68b0459e2841e0565cb6b45bc636f698bd/mqtt_as/mqtt_as.py#L154

So the code is (async) waiting for these events to be set elsewhere in the mqtt library.

**Reply by tferrin** [reply]
Well okay, then clearly I need to do a deeper dive into asyncio.Event() functionality. I guess I had superficially read the documentation where it says "asyncio.Event creates a new event which can be used to synchronize tasks" but didn't fully appreciate what this means. Will look for some tutorials. Thanks!

**Reply by peterhinch** [reply]
You might like to try [this tutorial](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md).

Re `mqtt_as` the module offers two alternative interfaces, traditional callbacks or `Event`-based. The choice is yours; some demos use callbacks, some use `Event`s. The choice determines what happens when an asynchronous event occurs, such as a message arriving. The callback concept is probably familiar to you. In the `Event` case the user runs a task which waits on the `Event` and then handles the message. In the `Event` case `mqtt_as` creates a queue: this avoids message loss in the case where messages might arrive in a burst whose rate briefly exceeds the speed at which the user task can process them.

In my opinion using tasks and `Event` instances is more in the spirit of `asyncio` than callbacks.

**Comment by tferrin** [comment]
OK. I got it now. The asyncio tutorial helped, as did a careful reading of section 3.5 Event based interface of the mqtt_as documentation. I originally got confused because when I looked at mqtt_as.py I didn't find either an "up" or "down" method, and it was, like, "how can this possibly work?"

And I'm all for the Event-based approach to mqtt_as. It was my desire to move away from callbacks that got me started down this path.

Thanks for the answers!
